controller/src/beerocks/master/wfa_ca.cpp

```diff
@@ -8,8 +16,91 @@
 
 #include <easylogging++.h>
 
+#include <algorithm>
+
 using namespace beerocks;
 
+std::string wfa_ca::wfa_ca_status(eWfaCaStatus status)
+{
+    switch (status) {
+    case RUNNING:
+        return "RUNNING";
+    case INVALID:
+        return "INVALID";
+    case ERROR:
+        return "ERROR";
+    case COMPLETE:
+        return "COMPLETE";
+    default:
+        break;
+    }
+    return std::string();
+}
+
+wfa_ca::eWfaCaCommand wfa_ca::wfa_ca_command(std::string command)
+{
+    std::transform(command.begin(), command.end(), command.begin(), ::toupper);
+
+    if (command == "CA_GET_VERSION") {
+        return CA_GET_VERSION;
+    } else if (command == "DEV_CONFIGURE_IE") {
+        return DEV_CONFIGURE_IE;
+    } else if (command == "DEV_GET_FRAME_INFO") {
+        return DEV_GET_FRAME_INFO;
+    } else if (command == "DEVICE_GET_INFO") {
+        return DEVICE_GET_INFO;
+    } else if (command == "DEV_EXEC_ACTION") {
+        return DEV_EXEC_ACTION;
+    } else if (command == "DEV_RESET_DEFAULT") {
+        return DEV_RESET_DEFAULT;
+    } else if (command == "DEV_SEND_1905") {
+        return DEV_SEND_1905;
+    } else if (command == "DEV_SEND_FRAME") {
+        return DEV_SEND_FRAME;
+    } else if (command == "DEV_SET_CONFIG") {
+        return DEV_SET_CONFIG;
+    }
+
+    return WFA_CA_COMMAND_MAX;
+}
+
+void wfa_ca::reply_invalid_error(Socket *sd, ieee1905_1::CmduMessageTx &cmdu_tx,
+                                 eWfaCaStatus status, const std::string &description)
+{
+    if (status != INVALID || status != ERROR) {
+        LOG(WARNING) << "function misuse, status is not INVALID or ERROR";
+        return;
+    }
+
+    std::string reply_string = "status," + wfa_ca_status(status) + ",errorCode," + description;
+    reply(sd, cmdu_tx, reply_string, false);
+}
+
+bool wfa_ca::reply(Socket *sd, ieee1905_1::CmduMessageTx &cmdu_tx, const std::string &reply_string,
+                   bool pending_second_response)
+{
+    auto response =
+        message_com::create_vs_message<beerocks_message::cACTION_BML_WFA_CA_CONTROLLER_RESPONSE>(
+            cmdu_tx);
+
+    if (!response) {
+        LOG(ERROR) << "Failed building cACTION_BML_WFA_CA_CONTROLLER_RESPONSE message!";
+        return false;
+    }
+
+    response->alloc_reply(reply_string.size() + 1);
+    auto reply_tuple = response->reply(0);
+    if (!std::get<0>(reply_tuple)) {
+        LOG(ERROR) << "getting reply entry has failed!";
+        return false;
+    }
+
+    char *reply_ptr = (char *)(&std::get<1>(reply_tuple));
+    string_utils::copy_string(reply_ptr, reply_string.c_str(), reply_string.size() + 1);
```

Comment:
Oh, so the 0 terminator *is* included then?

Reply:
it was